File: app/src/java/java/com/example/unityapp/views/MainActivity.java
package com.example.unityapp.views;

import android.os.Bundle;
import android.util.Log;

import androidx.activity.viewModels;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.practicemvvm.databinding.ActivityMainBinding;
import com.example.unityapp.data.Article;
import com.example.unityapp.utils.Filter;
import com.example.unityapp.utils.GenericFilter;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements MainListAdapter.OnItemClickListener {

    private ActivityMainBinding binding;
    private MyViewModel viewModel;

    private MainListAdapter mainListAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        viewModel = viewModels(MyViewModel.class).getValue();

        observeResponse();

        initUI();
    }

    private void initUI() {
        // RECYCLER
        binding.recView.setLayoutManager(new LinearLayoutManager(this));
        binding.recView.setHasFixedSize(true);
        mainListAdapter = new MainListAdapter(this);
        binding.recView.setAdapter(mainListAdapter);

        // CLICKS
        binding.btnFilterData.setOnClickListener(v -> viewModel.fetchData());
    }

    private void displayFilteredList(List<Article> articles) {
        List<Article> list = new ArrayList<>(articles);

        String filterStr = binding.etTextFilter.getText().toString();
        String filterNum = binding.etNumberFilter.getText().toString();

        if (!filterNum.isEmpty() && !filterStr.isEmpty()) {
            List<Filter<Object>> filters = new ArrayList<>();
            filters.add(new GenericFilter.NumberFilter(Integer.parseInt(filterNum)));
            filters.add(new GenericFilter.StringFilter(filterStr));
            list = GenericFilter.filterer(list, filters);
        }

        mainListAdapter.submitList(list);
    }

    private void observeResponse() {
        viewModel.getViewState().observe(this, response -> {
            displayFilteredList(response.getArticles());
        });
    }

    @Override
    public void onItemClick(Article article) {
        Log.d("", "");
    }
}

File: app/src/main/java/com/example/unityapp/views/MainActivity.kt
package com.example.unityapp.views

import android.os.Bundle
import android.util.Log
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.practicemvvm.databinding.ActivityMainBinding
import com.example.unityapp.data.Article
import com.example.unityapp.utils.Filter
import com.example.unityapp.utils.FilterTrials

class MainActivity : AppCompatActivity(), MainListAdapter.OnItemClickListener {

    private lateinit var binding: ActivityMainBinding
    private val viewModel by viewModels<MyViewModel>()

    private lateinit var mainListAdapter: MainListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        observeResponse()

        initUI()
    }


    private fun initUI() {
        //RECYCLER
        binding.recView.layoutManager = LinearLayoutManager(this)
        binding.recView.setHasFixedSize(true)
        mainListAdapter = MainListAdapter(this)
        binding.recView.adapter = mainListAdapter

        //CLICKS
        binding.btnFilterData.setOnClickListener {
            viewModel.fetchData()
        }
    }

    private fun displayFilteredList(articles: List<Article>) {
        var list = articles

        val filterStr = binding.etTextFilter.text
        val filterNum = binding.etNumberFilter.text

        if (filterNum.isNotBlank() && filterStr.isNotBlank()) {

            val filters = mutableListOf<Filter<Any>>(
                FilterTrials.NumberFilter(filterNum.toString().toInt()),
                FilterTrials.StringFilter(filterStr.toString())
            )
            list = FilterTrials.filterer(list, filters)
        }
        mainListAdapter.submitList(list)
    }

    private fun observeResponse() {
        viewModel.viewState.observe(this) { response ->
            displayFilteredList(response.articles)
        }
    }

    override fun onItemClick(note: Article) {
        Log.d("", "")
    }
}
